### ISSUE TITLE
Help running PX4 SITL

### ISSUE THREAD

**Issue by aks23021992** [mengchaoheng/dfgazebo#1]
Hello @mengchaoheng ,
 I went through your work and its really impressive to see your ducted fan UAV snapshot from gazebo. I am building very similar ducted fan UAV and wanted to first run SITL with PX4 firmware driving my setup. Can you provide detailed instructions on starting px4 SITL with your ducted fan sdf files?

**Comment by mengchaoheng** [comment]
@aks23021992 I'm glad to that I can help you. But,l don't know what your "went through" mean. Can you tell me your operation step by step? That will help me what you mean, since my repository is very much.

**Comment by aks23021992** [comment]
@mengchaoheng Sure I will brief my work first.
We are building ducted fan based UAV with 2 bldc motors for thrust + yaw control and 4 servo for fin control for pitch and roll actions. We also developed our own mixer similar to one shown on this repo. We are currently struggling to tune the aircraft in proper way. Also I didnt find any airframe support for ducted fan UAV in px4 airframes and neither any gazebo model for performing SITL of ducted fan elsewhere. That's where I came across your work and wanted to understand your followed procedure for tuning such vehicle on hardware as well as SITL setup using gazebo. 
Here are my queries :

Simulation queries :
1. In order to use your repo to fly ducted fan in SITL way on gazebo any steps which you guide me through? I can see liftdrag_plugin.cpp and header file in the repo. What is it's significance? Do I need to load them along with sdf?
2. Can I fly your ducted fan UAV with default parameter provided in repo in simulation? I do understand I need to put mixer file in px4 firmware at respective place. Will do the same.

Hardware tuning :
1. Can you help in getting started with tuning of such ducted fan vehicle over px4 firmware? How did you start tuning? We thought of using gimbal stand by locking one axis (say roll) and then tuning other axis (pitch). Any guidelines would be highly appreciated. Our aircraft is 30 cms in diameter and weights around 1.9 kg. Our thrust to weight ratio can takeoff the UAV easily which we have tested.

**Comment by mengchaoheng** [comment]
@aks23021992 

I actually forgot where I got this repository from, but from the documentation it works. What to do is very simple, you need to understand how px4 adds new models in advance.
summary as:
1. Write your sdf file like ductedfan4, then put it into PX4-Autopilot/Tools/sitl_gazebo/models, and then modify liftdrag_plugin.cpp to support the surface control.

2. Write parameter settings according to 15003_ducted_fan4 and add them to PX4-Autopilot/ROMFS/px4fmu_common/init.d/airframes/

3. Refer to ductedfan4.main.mix to write the mixing control settings and add them to /PX4-Autopilot/ROMFS/px4fmu_common/mixers-sitl/.

4. The compilation options need to be modified according to the px4 specification. For example, refer to the compilation settings of the plane, and add your new model in /PX4-Autopilot/platforms/posix/cmake/sitl_target.cmake. You just need to search the plane globally to see which files he has set up compilation options, then add your model. Remember to set up the cmakelists.txt file.

The setting of fmu-v5 is similar to the setting of sitl, and the setting of fmu-v5 allows you to run it on the actual aircraft.

In fact, it is normal for you to be confused, because this warehouse is not the warehouse that publishes all my work, it is not complete, I have other warehouses, but they are all private, and these private warehouses have all complete settings and can be launched with one click ductedfan UAV, but I'm not ready to open source it yet. If you are interested, you can continue to pay attention. Thanks!

**Comment by mengchaoheng** [comment]
@aks23021992  https://discuss.px4.io/t/flight-test-of-a-new-uav-ducted-fan-uav/27112

**Comment by mengchaoheng** [comment]
@aks23021992 
Can you tell me your institutional information? Which research institution or university are you from? I think we can keep communicating.

**Comment by aks23021992** [comment]
@mengchaoheng Thanks for all the information. I will try it out. I am from an university in India named Vellore institute of technology. 

Also can you provide stepwise guide on tuning this UAV on hardware?

**Comment by aks23021992** [comment]
@mengchaoheng 
While trying out SITL and compiling liftdrag_plugin.cpp I am facing the following error. Please let me know how to proceed.

`fatal error: Wind.pb.h: No such file or directory`

Wind.pb.h file is not found in current repo. Is there a modified version of Wind.pb.h to be included?

**Comment by mengchaoheng** [comment]
@aks23021992 I'm very sorry that I don't have time to write a tutorial. My suggestion is that you can check how they added and modified files when they added the model in the git commit log of px4. It is best to use vscode to see it. The idea of the liftdrag_plugin I modified is: add a few lines of code to make the force on the control surface related to the speed of the propeller, and then improve the shortcomings of the original liftdrag_plugin that the force of the control surface is only related to the speed.

**Comment by aks23021992** [comment]
@mengchaoheng Thanks for your response. Also apology for my late reply. I followed your told steps, added ductedfan model in my px4 firmware and also added mixer, compiled the code, also modified liftdrag_plugin code. I can see my mixer loaded with correct autostart code along with gazebo model properly spawning up.

When I arm the ductedfan from QGC, UAV just lifts off and runs away in z-axis (Altitude) in an uncontrolled manner. I doubt if my liftdrag_plugin is getting loaded correctly or not. Else can there be any other missing piece in the pipeline?

**Comment by mengchaoheng** [comment]
@aks23021992 you can debug it by print something. Maybe it was another liftdrag_plugin.cpp on your computer. And maybe your mix file is error. And maybe the parameters of .sdf is error.

**Comment by mengchaoheng** [comment]
@aks23021992 you can arm the uav and then switch to the angular rate mode, and then add the throttle but not takeoff, change the roll and pitch to test the fore of the control surface

**Comment by aks23021992** [comment]
@mengchaoheng I am loading liftdrag_plugin correctly. Tried printing something from plugin and got it printed on terminal.

When I try to arm UAV lifts up to 1 - 1.5 meters and topples. Ideally UAV shouldn't even lift while armed.

**Comment by mengchaoheng** [comment]
> @mengchaoheng I am loading liftdrag_plugin correctly. Tried printing something from plugin and got it printed on terminal. Also my mixer is loaded correctly as per terminal logs.
> 
> 
> 
> When I try to arm UAV lifts up to 1 - 1.5 meters and topples. Ideally UAV shouldn't even lift while armed.

@aks23021992 maybe the control channel of sdf file is error, change positive or negative

**Comment by aks23021992** [comment]
@mengchaoheng I modified control channel for roll and pitch servo drives signs and now UAV after ARM hovers straight without directly toppling but still isn't in control. Can you tell exactly which parameters in sdf to modify? I modified "input_scaling" parameter for servos.

**Comment by mengchaoheng** [comment]
> 

@aks23021992  can you send a video to here to show what happen？

**Comment by mengchaoheng** [comment]
@aks23021992 yes, it's `input_scaling`, if uav can be hover, then it should be control. if not, I don't know what heppen.

**Comment by mengchaoheng** [comment]
@aks23021992 An incomplete code is [here](https://github.com/Advanced-Ducted-Fan-Research/Ductedfan-Indi-RateControl), it just doesn't have the sitl_gazebo part modified. And this part only needs to change liftdrag_plugin.cpp and add sdf.

**Comment by aks23021992** [comment]
https://user-images.githubusercontent.com/105961319/209668286-75c78ba6-f8f2-4d79-9469-3803f0ca3a04.mp4

Please have look at video attached to see behaviour of UAV after takeoff from QGround control.
Also my PX4 firmware version is v1.11.3-0.2.3. Will that have any problem?

**Comment by mengchaoheng** [comment]
you can test ductedfan4. Or I will push a new code to github, just wait a few days.

**Comment by aks23021992** [comment]
@mengchaoheng Waiting for your code push.
In the meanwhile I will test ductedfan4 as well.
Parameters for ductedfan4 are to be taken from your PX4 codebased mentioned [over here](https://github.com/mengchaoheng/PX4-Autopilot/tree/02295f65d04f42ae54234652278c1f42b389edc1/ROMFS/px4fmu_common/init.d).
